internal/controller/operator/factory/vmagent/vmagent.go

```diff
@@ -146,125 +142,38 @@ func CreateOrUpdate(ctx context.Context, cr *vmv1beta1.VMAgent, rclient client.C
-		if prevObjectSpec != nil {
-			prevAppObject, ok := prevObjectSpec.(*appsv1.StatefulSet)
-			if ok {
-				prevSTS = prevAppObject
-				prevSTS, err = k8stools.RenderPlaceholders(prevSTS, defaultPlaceholders)
-				if err != nil {
-					return fmt.Errorf("cannot fill placeholders for prev sts in vmagent: %w", err)
-				}
-			}
-		}
-		newDeploy, err = k8stools.RenderPlaceholders(newDeploy, defaultPlaceholders)
+	if prevCR != nil {
+		prevAppTpl, err = newK8sApp(prevCR, ac)
 		if err != nil {
-			return fmt.Errorf("cannot fill placeholders for sts in vmagent: %w", err)
-		}
-		stsOpts := reconcile.STSOptions{
-			HasClaim:       len(newDeploy.Spec.VolumeClaimTemplates) > 0,
-			SelectorLabels: cr.SelectorLabels,
-		}
-		if err := reconcile.HandleSTSUpdate(ctx, rclient, stsOpts, newDeploy, prevSTS); err != nil {
-			return err
-		}
-		stsNames[newDeploy.Name] = struct{}{}
-	case *appsv1.DaemonSet:
-		var prevDeploy *appsv1.DaemonSet
-		if prevObjectSpec != nil {
-			prevAppObject, ok := prevObjectSpec.(*appsv1.DaemonSet)
-			if ok {
-				prevDeploy = prevAppObject
-			}
+			return fmt.Errorf("cannot build new deploy for vmagent: %w", err)
 		}
-		if err := reconcile.DaemonSet(ctx, rclient, newDeploy, prevDeploy); err != nil {
-			return err
+		if reflect.TypeOf(newAppTpl) != reflect.TypeOf(prevAppTpl) {
+			prevAppTpl = nil
 		}
-	default:
-		panic(fmt.Sprintf("BUG: unexpected deploy object type: %T", newDeploy))
-	}
-	if err := finalize.RemoveOrphanedDeployments(ctx, rclient, cr, deploymentNames); err != nil {
-		return err
-	}
-	if err := finalize.RemoveOrphanedSTSs(ctx, rclient, cr, stsNames); err != nil {
-		return err
 	}
-	if err := removeStaleDaemonSet(ctx, rclient, cr); err != nil {
-		return fmt.Errorf("cannot remove vmagent daemonSet: %w", err)
-	}
-	return nil
-}
 
-func createOrUpdateShardedDeploy(ctx context.Context, rclient client.Client, cr, prevCR *vmv1beta1.VMAgent, newDeploy, prevDeploy runtime.Object) error {
-	deploymentNames := make(map[string]struct{})
-	stsNames := make(map[string]struct{})
-	shardsCount := *cr.Spec.ShardCount
-	logger.WithContext(ctx).Info(fmt.Sprintf("using cluster version of VMAgent with shards count=%d", shardsCount))
+	deploymentToKeep := make(map[string]struct{})
+	stsToKeep := make(map[string]struct{})
+	pdbToKeep := make(map[string]struct{})
+	shardCount := cr.GetShardCount()
+	prevShardCount := prevCR.GetShardCount()
+	logger.WithContext(ctx).Info(fmt.Sprintf("using cluster version of VMAgent with shards count=%d", shardCount))
 
 	isUpscaling := false
-	if prevCR != nil && prevCR.Spec.ShardCount != nil {
-		if *prevCR.Spec.ShardCount < shardsCount {
-			logger.WithContext(ctx).Info(fmt.Sprintf("VMAgent shard upscaling from=%d to=%d", *prevCR.Spec.ShardCount, shardsCount))
-			isUpscaling = true
-		}
+	if prevShardCount > 1 && prevShardCount < shardCount {
+		logger.WithContext(ctx).Info(fmt.Sprintf("VMAgent shard upscaling from=%d to=%d", prevShardCount, shardCount))
```

Comment:
```suggestion
		logger.WithContext(ctx).Info(fmt.Sprintf("VMAgent shard upscaling", "from", prevShardCount, "to", shardCount))
```
nitpick: structured logging